runtime/lua/vim/lsp/_watchfiles.lua

```diff
@@ -0,0 +1,346 @@
+---
+---@param pattern string|table The glob pattern (raw or parsed) to match.
+---@param s string The string to match against pattern.
+---@return bool Whether or not pattern matches s.
+local function match(pattern, s)
+  if type(pattern) == "string" then
+    pattern = parse(pattern)
+  end
+  -- Since Lua's built-in string pattern matching does not have an alternate
+  -- operator like '|', `parse` will construct one pattern for each possible
+  -- alternative. Any pattern that matches thus matches the glob.
+  for _, p in ipairs(pattern) do
+    if s:match("^"..p.."$") then return true end
+  end
+  return false
+end
+
+---@private
+--- Joins filepath elements by platform-specific separator.
+---
+---@param args ... The path elements. The first element must be absolute.
+---@return string The joined path.
+local function filepath_join(...)
+  local dir = ...
+  return table.concat({...}, dir:match('^([a-zA-Z]:)(.*)') and "\\" or "/")
+end
+
+-- Cache of libuv handles per directory, per LSP client, per registration ID.
+local watched_dirs = {}
+
+local change_queue = {}
+local change_cache = {}
+local queue_timers = {}
+
+---@private
+--- Creates callbacks invoked on watched file events.
+---
+---@param dir string Absolute path to the watched directory.
+---@param client_id number The LSP client's ID.
+---@param reg_id string The ID used to register the request.
+---@return function The callback invoked on watched file events.
+local function get_callback(dir, client_id, reg_id)
+  return function(filename, events)
+    local path = filepath_join(dir, filename)
+
+    local type = events.change and protocol.FileChangeType.Changed or 0
+    if events.rename then
+      local _, err, errname = uv.fs_stat(path)
+      if errname == "ENOENT" then
+        type = protocol.FileChangeType.Deleted
+      else
+        assert(not err, err)
+        type = protocol.FileChangeType.Created
+      end
+    end
+
+    local matches_filter = false
+    local filters = watched_dirs[dir].callbacks[client_id][reg_id].filters
+    for _, filter in ipairs(filters) do
+      if match(filter.pattern, path) and math.floor(filter.kind / (2^(type-1))) % 2 == 1 then
+        matches_filter = true
+        break
+      end
+    end
+    if not matches_filter then return end
+
+    local change = {
+      uri = vim.uri_from_fname(path),
+      type = type,
+    }
+    change_cache[client_id] = change_cache[client_id] or {}
+    local last_type = change_cache[client_id][change.uri]
+    if not last_type or last_type ~= change.type then
+      change_queue[client_id] = change_queue[client_id] or {}
+      table.insert(change_queue[client_id], change)
+      change_cache[client_id][change.uri] = change.type
+    end
+    if not queue_timers[client_id] then
+      local t = uv.new_timer()
+      t:start(100, 0, function()
```

Comment:
This function run after the timer goes off sends the LSP notification with the current batch of unsent changes seen. I did try wrapping this callback with `vim.schedule_wrap`, but that didn't seem to send the notifications right when the timer was set to go off the way I expected. Does that raise any red flags?

Reply:
`schedule_wrap` creates a new function that would still need to be called. There's `vim.schedule` to defer execution to the main loop. Or if it can run in the luv loop you could use `vim.defer_fn`, which is a shortcut for the timer pattern you use here.